test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -105,141 +109,277 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
-		}()
+		l.cleanupSteps = append(l.cleanupSteps, func() {
+			framework.Logf("deleting storage class %s", l.sc.Name)
+			framework.ExpectNoError(l.cs.StorageV1().StorageClasses().Delete(context.TODO(), l.sc.Name, metav1.DeleteOptions{}))
+		})
 
 		ginkgo.By("creating a claim")
-		pvc, err = cs.CoreV1().PersistentVolumeClaims(pvc.Namespace).Create(context.TODO(), pvc, metav1.CreateOptions{})
+		l.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.pvc.Namespace).Create(context.TODO(), l.pvc, metav1.CreateOptions{})
 		framework.ExpectNoError(err)
-		defer func() {
-			framework.Logf("deleting claim %q/%q", pvc.Namespace, pvc.Name)
+		l.cleanupSteps = append(l.cleanupSteps, func() {
+			framework.Logf("deleting claim %q/%q", l.pvc.Namespace, l.pvc.Name)
 			// typically this claim has already been deleted
-			err = cs.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(context.TODO(), pvc.Name, metav1.DeleteOptions{})
+			err = l.cs.CoreV1().PersistentVolumeClaims(l.pvc.Namespace).Delete(context.TODO(), l.pvc.Name, metav1.DeleteOptions{})
 			if err != nil && !apierrors.IsNotFound(err) {
-				framework.Failf("Error deleting claim %q. Error: %v", pvc.Name, err)
+				framework.Failf("Error deleting claim %q. Error: %v", l.pvc.Name, err)
 			}
-		}()
+		})
 
 		ginkgo.By("starting a pod to use the claim")
 		command := "echo 'hello world' > /mnt/test/data"
-		pod := StartInPodWithVolume(cs, pvc.Namespace, pvc.Name, "pvc-snapshottable-tester", command, config.ClientNodeSelection)
-		defer StopPod(cs, pod)
+		l.pod = StartInPodWithVolume(l.cs, l.pvc.Namespace, l.pvc.Name, "pvc-snapshottable-tester", command, config.ClientNodeSelection)
+		l.cleanupSteps = append(l.cleanupSteps, func() {
+			StopPod(l.cs, l.pod)
+		})
 
-		err = e2epv.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, cs, pvc.Namespace, pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
+		err = e2epv.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, l.cs, l.pvc.Namespace, l.pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
 		framework.ExpectNoError(err)
 
 		ginkgo.By("checking the claim")
 		// Get new copy of the claim
-		pvc, err = cs.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(context.TODO(), pvc.Name, metav1.GetOptions{})
+		l.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.pvc.Namespace).Get(context.TODO(), l.pvc.Name, metav1.GetOptions{})
 		framework.ExpectNoError(err)
 
 		// Get the bound PV
-		_, err = cs.CoreV1().PersistentVolumes().Get(context.TODO(), pvc.Spec.VolumeName, metav1.GetOptions{})
+		_, err = l.cs.CoreV1().PersistentVolumes().Get(context.TODO(), l.pvc.Spec.VolumeName, metav1.GetOptions{})
 		framework.ExpectNoError(err)
+	}
+	cleanup := func(l *snapshottableLocal) {
+		// Depending on how far the test executed, cleanup accordingly
+		// Execute in reverse order, similar to defer stack
+		err := tryFunc(func() {
+			for i := len(l.cleanupSteps) - 1; i >= 0; i-- {
+				l.cleanupSteps[i]()
```

Comment:
nit: Should this be tryFunc(l.cleanupSteps[i]()), so that we continue trying cleanup of other resources?

Reply:
Ah, I misunderstood tryFunc and recover(). Since it's one recovered per tryFunc, I added a tryFunc per step.